bindings/rust/src/bytes.rs

```diff
@@ -0,0 +1,55 @@
+use alloc::{borrow::ToOwned, vec::Vec};
+use core::{
+    convert::From,
+    ops::{Bound, Deref, RangeBounds},
+};
+
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct Bytes(Vec<u8>);
+
+impl From<Vec<u8>> for Bytes {
+    fn from(value: Vec<u8>) -> Self {
+        Self(value)
+    }
+}
+
+impl From<&[u8]> for Bytes {
+    fn from(value: &[u8]) -> Self {
+        Self(value.to_owned())
+    }
+}
+
+impl From<Bytes> for Vec<u8> {
+    fn from(value: Bytes) -> Self {
+        value.0
+    }
+}
+
+impl Deref for Bytes {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Bytes {
+    pub fn from_static(bytes: &[u8]) -> Self {
+        Self::from(bytes)
+    }
+
+    pub fn slice(&self, range: impl RangeBounds<usize>) -> Self {
+        let len = self.len();
+        let begin = match range.start_bound() {
+            Bound::Included(&n) => n,
+            Bound::Excluded(&n) => n + 1,
+            Bound::Unbounded => 0,
+        };
+        let end = match range.end_bound() {
+            Bound::Included(&n) => n + 1,
```

Comment:
I think it should be `n`

Reply:
I think the current code is mostly correct. But there's an edge case, when `n` is `usize::MAX`. It might be better to pass the range to `self.0[..].index` directly.

Reply:
Sorry, I misunderstood. It's correct.

```rust
use std::ops::Bound;
use std::ops::RangeBounds;
fn main() {
    let a = vec![0, 1, 2, 3, 4, 5];

    let exclude_2 = slice(a.clone(), 0..2);
    let include_2 = slice(a.clone(), 0..=2);

    let exclude_5 = slice(a.clone(), 0..5);
    let include_5 = slice(a.clone(), 0..=5);

    dbg!(&exclude_2);
    dbg!(&include_2);

    dbg!(&exclude_5);
    dbg!(&include_5);
}

pub fn slice(src: Vec<u8>, range: impl RangeBounds<usize>) ->Vec<u8> {
    let len = src.len();
    let begin = match range.start_bound() {
        Bound::Included(&n) => n,
        Bound::Excluded(&n) => n + 1,
        Bound::Unbounded => 0,
    };
    let end = match range.end_bound() {
        Bound::Included(&n) => n + 1,
        Bound::Excluded(&n) => n,
        Bound::Unbounded => len,
    };
    src[begin..end].to_vec()
}
```

```bash
 cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/rust-playground`
[src/main.rs:21] &exclude_2 = [
    0,
    1,
]
[src/main.rs:22] &include_2 = [
    0,
    1,
    2,
]
[src/main.rs:24] &exclude_5 = [
    0,
    1,
    2,
    3,
    4,
]
[src/main.rs:25] &include_5 = [
    0,
    1,
    2,
    3,
    4,
    5,
]

```

Reply:
Use [SliceIndex for (Bound<usize>, Bound<usize>)](https://doc.rust-lang.org/stable/core/slice/trait.SliceIndex.html#impl-SliceIndex%3C%5BT%5D%3E-for-(Bound%3Cusize%3E,+Bound%3Cusize%3E))

```rust
let range = (range.start_bound.cloned(), range.end_bound().cloned());
self.0.index(range)
```

Reply:
> Use [SliceIndex for (Bound, Bound)](https://doc.rust-lang.org/stable/core/slice/trait.SliceIndex.html#impl-SliceIndex%3C%5BT%5D%3E-for-(Bound%3Cusize%3E,+Bound%3Cusize%3E))
> 
> ```rust
> let range = (range.start_bound.cloned(), range.end_bound().cloned());
> self.0.index(range)
> ```

But the SliceIndex trait is unsafe.

Reply:
It's safe to use. It the same as `self.0[range]`. (Actually, `self.0[range]` is better.)